[Opt Overlap] Implement register_fused_optim in DDP

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71620
* #71608
* #71607
* __->__ #71606
* #71605
* #71604
* #71603
* #71602

Implements a `_register_fused_optim` function in DDP that allows users to register an optimizer to run in step with the allreduce. The function takes in a optimizer class, args and kwargs and calls into the `_OverlappedOptimizer` built below to construct the appropriate functional optimizer and eventually register it as a communication hook.

Current usage and constraints are documented in the function, and https://github.com/pytorch/pytorch/issues/71595 is filed to track use cases that are currently unsupported.

Example of usage:

```
ddp_model_with_optimizer_hook._register_fused_optim(
                        optim_cls,
                        *functional_optim_args,
                        **functional_optim_kwargs,
                    )
```
Differential Revision: [D33694037](https://our.internmc.facebook.com/intern/diff/D33694037/)

rohan-varma committed:
[Opt Overlap] Implement register_fused_optim in DDP

Per title

Differential Revision: [D33694037](https://our.internmc.facebook.com/intern/diff/D33694037/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Opt Overlap] Implement register_fused_optim in DDP"


Implements a `_register_fused_optim` function in DDP that allows users to register an optimizer to run in step with the allreduce. The function takes in a optimizer class, args and kwargs and calls into the `_OverlappedOptimizer` built below to construct the appropriate functional optimizer and eventually register it as a communication hook.

Current usage and constraints are documented in the function, and https://github.com/pytorch/pytorch/issues/71595 is filed to track use cases that are currently unsupported.

Example of usage:

```
ddp_model_with_optimizer_hook._register_fused_optim(
                        optim_cls,
                        *functional_optim_args,
                        **functional_optim_kwargs,
                    )
```
Differential Revision: [D33694037](https://our.internmc.facebook.com/intern/diff/D33694037/)

[ghstack-poisoned]

cbalioglu commented on torch/nn/parallel/distributed.py:
```diff
@@ -1435,6 +1435,65 @@ def _register_builtin_comm_hook(self, comm_hook_type):
+            overlapped_optim.register_ddp(self)
+        except NotImplementedError:
+            raise RuntimeError(
+                f"{optim} does not support overlapped DDP. Please file an issue to PyTorch."
```
Nit: In case we aim to support 3rd party `OverlappedOptimizer`s, we might reword this sentence a bit like:

```
{optim} does not support overlapped DDP. Please file an issue to PyTorch or to the respective owner of {optim}.
```

rohan-varma replied:
Added in the latest update!

cbalioglu reviewed (approved):
Just a nit comment again. Looks good to me.

rohan-varma committed:
Update on "[Opt Overlap] Implement register_fused_optim in DDP"


Implements a `_register_fused_optim` function in DDP that allows users to register an optimizer to run in step with the allreduce. The function takes in a optimizer class, args and kwargs and calls into the `_OverlappedOptimizer` built below to construct the appropriate functional optimizer and eventually register it as a communication hook.

Current usage and constraints are documented in the function, and https://github.com/pytorch/pytorch/issues/71595 is filed to track use cases that are currently unsupported.

Example of usage:

```
ddp_model_with_optimizer_hook._register_fused_optim(
                        optim_cls,
                        *functional_optim_args,
                        **functional_optim_kwargs,
                    )
```
Differential Revision: [D33694037](https://our.internmc.facebook.com/intern/diff/D33694037/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Opt Overlap] Implement register_fused_optim in DDP"


Implements a `_register_fused_optim` function in DDP that allows users to register an optimizer to run in step with the allreduce. The function takes in a optimizer class, args and kwargs and calls into the `_OverlappedOptimizer` built below to construct the appropriate functional optimizer and eventually register it as a communication hook.

Current usage and constraints are documented in the function, and https://github.com/pytorch/pytorch/issues/71595 is filed to track use cases that are currently unsupported.

Example of usage:

```
ddp_model_with_optimizer_hook._register_fused_optim(
                        optim_cls,
                        *functional_optim_args,
                        **functional_optim_kwargs,
                    )
```
Differential Revision: [D33694037](https://our.internmc.facebook.com/intern/diff/D33694037/)

[ghstack-poisoned]
